diff --git a/Makefile b/Makefile
index b7c176d..ebd3435 100644
--- a/Makefile
+++ b/Makefile
@@ -1,15 +1,14 @@
 CXX=g++ -std=c++17
-RM=rm -f
-RMDIR=rm -rf
+OBJDIR=obj
+LIBDIR?=lib
+PREFIX=/usr
 HFILES:=$(shell find src/ -type f -iname *.h -print)
 CXXFILES:=$(shell find src/ -type f -iname *.cpp -print)
-GTKFLAGS=$(shell pkg-config gtkmm-3.0 --cflags --libs)
-CXXFLAGS=$(GTKFLAGS) -Wall -lsteam_api -lcurl -lyajl -ldl
-LDFLAGS=-L${CURDIR}/bin
-OBJDIR=obj
+GTKFLAGS:=$(shell pkg-config gtkmm-3.0 --cflags --libs)
+CXXFLAGS+=$(GTKFLAGS) -Wall -lsteam_api -lcurl -lyajl -ldl
+LDFLAGS+=-L${CURDIR}/bin
 OBJS=$(addprefix ${OBJDIR}/,$(subst .cpp,.o,${CXXFILES}))
 
-all: CXXFLAGS += -O3
 all: ${CURDIR}/bin/samrewritten
 	@echo -e "==== Use '\033[1mmake dev\033[0m' to keep debug symbols"
 	@echo -e "==== Use '\033[1mmake clean\033[0m' to remove object files"
@@ -18,8 +17,23 @@ all: ${CURDIR}/bin/samrewritten
 dev: CXXFLAGS += -g -DDEBUG_CERR
 dev: ${CURDIR}/bin/samrewritten
 
+.PHONY: clean
 clean:
-	${RMDIR} ${OBJDIR}
+	rm -rf ${OBJDIR}
+
+.PHONY: install
+install: bin/launch.sh bin/samrewritten bin/libsteam_api.so
+	mkdir -p ${DESTDIR}${PREFIX}/${LIBDIR}/SamRewritten/{bin,glade,assets}
+	mkdir -p ${DESTDIR}${PREFIX}/share/icons/hicolor/{64x64,256x256}/apps
+	mkdir -p ${DESTDIR}${PREFIX}/share/applications
+	mkdir -p ${DESTDIR}${PREFIX}/bin
+	cp $^ ${DESTDIR}${PREFIX}/${LIBDIR}/SamRewritten/bin/
+	ln -s ${DESTDIR}${PREFIX}/${LIBDIR}/SamRewritten/bin/launch.sh ${DESTDIR}${PREFIX}/bin/samrewritten
+	cp glade/main_window.glade ${DESTDIR}${PREFIX}/${LIBDIR}/SamRewritten/glade/main_window.glade
+	cp assets/icon_64.png ${DESTDIR}${PREFIX}/share/icons/hicolor/64x64/apps/samrewritten.png
+	cp assets/icon_256.png ${DESTDIR}${PREFIX}/share/icons/hicolor/256x256/apps/samrewritten.png
+	cp assets/icon_256.png ${DESTDIR}${PREFIX}/${LIBDIR}/SamRewritten/assets/
+	cp package/samrewritten.desktop ${DESTDIR}${PREFIX}/share/applications/
 
 ${CURDIR}/bin/samrewritten: $(OBJS)
 	${CXX} -o ${CURDIR}/bin/samrewritten $(OBJS) ${LDFLAGS} ${CXXFLAGS}
